[network][service-with-healthchecks] New internal LB with per-port healthchecks (#9371)
deckhouse/deckhouse: ee/modules/610-service-with-healthchecks/docs/README.md

---
title: "The service-with-healthchecks module"
---

> In kubernetes, traffic is directed only to "live" Pods (using Service and EndpointSlices), while "liveliness" is determined by only one for the whole under test - readiness-probe of the pod.

Reviewer: ```suggestion
In kubernetes, traffic is directed only to "ready" Pods (using Service and EndpointSlices), while "readiness" is determined by only one for the whole under test - readiness-probe of the pod.
```

There are options when the pods contain several processes (like VMs) and each of them can handle traffic independently. In this case, one readiness-probe is not enough.

> This module implements the ability to independently check the availability of each process and, based on the results of a set of new probes, direct traffic to them.

Reviewer: ```suggestion
This module implements the ability to independently check the availability of each process and, based on the results of a new set of probes, direct traffic to them.
```

You can set up a new balancing method using the ServiceWithHealthchecks resource:

> - its specification is identical to Service with the addition of the healthchecks section, which contains a set of new probes.

Reviewer: ```suggestion
- its specification is identical to Service with the addition of the healthchecks section, which contains a new set of probes.
```

> - when creating this resource, child services (without selector) and corresponding endpointslices are created.

Reviewer: ```suggestion
- when creating this resource, child Services (without selector) and corresponding EndpointSlices are created.
```

- **Configuration**: A readiness probe is configured through the pod specification. It can be implemented as an HTTP request, TCP connection, or command-line check (exec). For example, an HTTP probe sends a request to a specific URL, and if the response is successful (usually a 200 code), the pod is considered "ready."

- **Behavior**: If the readiness probe fails, the pod will be marked as "not ready," and all requests to the service will be redirected to other "live" pods. Once the pod successfully passes the check, it will be available for traffic again.

- **Importance**: The use of readiness probes is critical for maintaining high availability and optimal load distribution among pods. This is especially important in distributed systems where the state of applications can change dynamically.

Thus, readiness probes play a key role in traffic management in Kubernetes — requests are processed only by available pods.

There are scenarios where pods contain multiple processes, each capable of handling traffic independently, similar to virtual machines. In such cases, a single readiness probe is insufficient to determine the availability of all processes.

> The service-with-healthchecks module provides the ability to independently check the availability of each process and, based on the results of the new set of probes, direct traffic to the appropriate instances. This is useful, for example, for complex multi-container applications in Kubernetes where a "common" health check is defined for multiple containers in a pod.

Reviewer 2: ```suggestion
The service-with-healthchecks module provides the ability to independently check the availability of more than one process and, based on the results of the new set of probes, direct traffic to the appropriate instances. This is useful, for example, for complex multi-container applications in Kubernetes where a "common" health check is defined for multiple containers in a pod.
```

> A new load-balancing method can be configured using the `ServiceWithHealthchecks` resource:

Reviewer 2: ```suggestion
A new load-balancing method can be configured using the [ServiceWithHealthchecks](cr.html#servicewithhealthchecks) resource:
```

---
title: "The service-with-healthchecks module"
description: "Readiness Probes with the service-with-healthchecks Module in Deckhouse Kubernetes Platform"
---

In Kubernetes, traffic is only routed to pods that are up and running using `Service` and `EndpointSlice` resources. This ensures high availability of the application, as requests are only handled by the pods that are ready to receive them.

Kubernetes finds out if a pod is available and ready to receive traffic by using a check known as a readiness probe.

### What is a readiness probe?

- **Purpose**: The primary goal of a readiness probe is to ensure that the application running in the pod is ready to process requests. If the probe fails, Kubernetes will not route traffic to that pod. This helps to avoid errors and increases the application stability.

- **Configuration**: A readiness probe is configured via the pod specification. It can take the form of an HTTP request, an attempt to establish a TCP connection, or a command line (exec) check. For example, an HTTP probe sends a request to a specific URL, and if the response is successful (usually HTTP code 200 OK), the pod is considered "ready".

> - **Behavior**: If the readiness probe fails, the pod will be marked as "not ready," and all requests to the service will be reruouted to other running pods. Once the pod successfully passes the check, it will be able to receive traffic again.

Reviewer 2: ```suggestion
- **Behavior**: If the readiness probe fails, the pod will be marked as "not ready," and all requests to the service will be rerouted to other running pods. Once the pod successfully passes the check, it will be able to receive traffic again.
```

---
title: "The service-with-healthchecks module"
---

> Options when pods contain many processes (like VMs) and each of them can handle traffic independently. For this purpose, it is necessary to be able to evaluate the ability of these processes to accept incoming traffic. For this purpose we have implemented an extended set in the new ServiceWithHealthchecks resource. Implementation of the module with an extended version of the standard Service resource. The specification of the ServiceWithHealthchecks resource is extended with a set of healthchecks in comparison with the standard one. At this stage http and tcp probes are implemented.

Reviewer: It is necessary to improve the general description and add a brief explanation of what components the module consists of and how they interact.

## ServiceWithHealthcheck load balancer capabilities

Unlike the regular load balancer, in which readiness probes are tied to the container state, `ServiceWithHealthcheck` allows you to set up active probes on individual TCP ports. In this way, each of the load balancers dealing with the same pod can operate independently of the others.

You can configure this balancing method using the [ServiceWithHealthchecks](cr.html#servicewithhealthchecks) resource:

* Its specification is the same as the regular `Service` except for the `healthcheck` section, which contains a set of probes.
* Currently, three types of probes are supported:
  * `TCP` — a regular probe that establishes a TCP connection.
  * `HTTP` — a probe that sends an HTTP request and waits for a specific response code.
  * `PostgreSQL` — a probe that sends an SQL query and waits for it to complete successfully.

Examples can be found in the [documentation](examples.html).

## How the ServiceWithHealthcheck load balancer works

The load balancer is made up of two components:

* The **controller** runs on the cluster master nodes and manages `ServiceWithHealthcheck` resources,

> * The agents operate on each cluster node and carry out probing for pods that run on the local node.

Reviewer 3: ```suggestion
* The agents operate on each cluster node and carry out probing for pods that run on that node.
```

Балансировщик ServiceWithHealthcheck спроектирован так, чтобы не зависеть от реализации CNI, используя при этом стандартные ресурсы `Service` и `EndpointSlice`:

> * Контроллер при создании ресурса `ServiceWithHealthcheck` автоматически создает одноименный ресурс Service в том же пространстве имен с пустым полем `selector`. Это позволяет избежать создания `EndpointSlice` стандартным контроллером, что устраняет привязку к конкретным адресам для балансировки запросов.

Reviewer 3: ```suggestion
* Контроллер при создании ресурса `ServiceWithHealthcheck` автоматически создает одноименный ресурс Service в том же пространстве имен с пустым полем `selector`. Это позволяет избежать создания стандартным контроллером `EndpointSlice`, которые используются для настройки балансировки.
```
